File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool sqr(vector<vector<char>>& board, int rowStart, int colStart, unordered_map<int,unordered_set<int>>&c, unordered_map<int,unordered_set<int>>&r) {
        unordered_set<int>s;
        for(int i=rowStart; i<3+rowStart; i++) {
            for(int j=colStart; j < colStart+3; j++) {
                if(board[i][j] != '.') {
                    if(s.find(board[i][j]) == s.end()) {
                        s.insert(board[i][j]);
                    } else {
                        return false;
                    }
                    if(c[j].find(board[i][j]) == c[j].end()) {
                        c[j].insert(board[i][j]);
                    } else {
                        return false;
                    }
                    if(r[i].find(board[i][j]) == r[i].end()) {
                        r[i].insert(board[i][j]);
                    } else {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int,unordered_set<int>>c;
        unordered_map<int,unordered_set<int>>r;
        for(int i=0; i<9; i=i+3) {
            for(int j=0; j<9; j=j+3) {
                if(!sqr(board,i,j,c,r)) return false;
            }
        }
        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        r = {}
        c = {}
        
        for i in range(0,9):
            r[i] = set()
            c [i] = set()
        
        for i in range(3):
            for j in range(3):
                s = set()
                for k in range(3):
                    for l in range(3):
                        if board[i*3+k][j*3+l] == '.':
                            continue
                        if board[i*3+k][j*3+l] in s:
                            return False
                        else:
                            s.add(board[i*3+k][j*3+l])
                            
                            if board[i*3+k][j*3+l] in r[i*3+k]:
                                return False
                            else:
                                r[i*3+k].add(board[i*3+k][j*3+l])
                            
                            if board[i*3+k][j*3+l] in c[j*3+l]:
                                return False
                            else:
                                c[j*3+l].add(board[i*3+k][j*3+l])         
        
        return True
                